File: niro/src/main/java/nl/uva/se/sc/niro/parser/ErrorListener.java
package nl.uva.se.sc.niro.parser;

import org.antlr.v4.runtime.ANTLRErrorListener;
import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;
import org.antlr.v4.runtime.atn.ATNConfigSet;
import org.antlr.v4.runtime.dfa.DFA;
import org.antlr.v4.runtime.misc.Interval;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ql.QLParser;

import java.util.BitSet;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class ErrorListener implements ANTLRErrorListener {
    private static final Logger LOGGER = LogManager.getLogger();
    private List<ParseErrorInfo> parseErrors = new LinkedList<>();

	@Override
	public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine,
			String msg, RecognitionException e) {

		List<String> stack = ((Parser) recognizer).getRuleInvocationStack();
		Collections.reverse(stack);
		parseErrors.add(new SyntaxErrorInfo(line, charPositionInLine, offendingSymbol.toString(), msg, stack, e));
	}

	@Override
	public void reportAmbiguity(Parser recognizer, DFA dfa, int startIndex, int stopIndex, boolean exact, BitSet ambigAlts, ATNConfigSet configs) {
        QLParser parser = (QLParser) recognizer;
        Interval interval = Interval.of(startIndex, stopIndex);
        LOGGER.info("Ambiguous construct around ["+parser.getTokenStream().getText(interval) + "].");
	}

	@Override
	public void reportAttemptingFullContext(Parser recognizer, DFA dfa, int startIndex, int stopIndex, BitSet conflictingAlts, ATNConfigSet configs) {
        QLParser parser = (QLParser) recognizer;
        Interval interval = Interval.of(startIndex, stopIndex);
        LOGGER.info("Need full context on ["+parser.getTokenStream().getText(interval) + "] to make decision.");
	}

	@Override
	public void reportContextSensitivity(Parser recognizer, DFA dfa, int startIndex, int stopIndex, int prediction, ATNConfigSet configs) {
        QLParser parser = (QLParser) recognizer;
        Interval interval = Interval.of(startIndex, stopIndex);
        LOGGER.info("Context sensitive construct while dealing with ["+parser.getTokenStream().getText(interval) + "].");
	}

	public List<ParseErrorInfo> getParseErrors() {
		return parseErrors;
	}
}

File: niro/src/main/scala/nl/uva/se/sc/niro/parser/ErrorListener.scala
package nl.uva.se.sc.niro.parser

import java.util

import nl.uva.se.sc.niro.errors.Errors._
import org.antlr.v4.runtime.atn.ATNConfigSet
import org.antlr.v4.runtime.dfa.DFA
import org.antlr.v4.runtime.misc.Interval
import org.antlr.v4.runtime.{ ANTLRErrorListener, Parser, RecognitionException, Recognizer }
import org.apache.logging.log4j.scala.Logging

import scala.collection.mutable.ListBuffer

class ErrorListener extends Logging with ANTLRErrorListener {
  var parseErrors = new ListBuffer[Error]

  override def syntaxError(
      recognizer: Recognizer[_, _],
      offendingSymbol: Any,
      line: Int,
      charPositionInLine: Int,
      msg: String,
      e: RecognitionException): Unit = {
    parseErrors += SyntaxErrorInfo(line, charPositionInLine, offendingSymbol.toString, msg, e)
  }

  override def reportAmbiguity(
      parser: Parser,
      dfa: DFA,
      startIndex: Int,
      stopIndex: Int,
      exact: Boolean,
      ambigAlts: util.BitSet,
      configs: ATNConfigSet): Unit = {
    val interval = Interval.of(startIndex, stopIndex)
    logger.info(s"Ambiguous construct around [${parser.getTokenStream.getText(interval)}].")
  }

  override def reportAttemptingFullContext(
      parser: Parser,
      dfa: DFA,
      startIndex: Int,
      stopIndex: Int,
      conflictingAlts: util.BitSet,
      configs: ATNConfigSet): Unit = {
    val interval = Interval.of(startIndex, stopIndex)
    logger.info(s"Need full context on [${parser.getTokenStream.getText(interval)}] to make decision.")
  }

  override def reportContextSensitivity(
      parser: Parser,
      dfa: DFA,
      startIndex: Int,
      stopIndex: Int,
      prediction: Int,
      configs: ATNConfigSet): Unit = {
    val interval = Interval.of(startIndex, stopIndex)
    logger.info(s"Context sensitive construct while dealing with [${parser.getTokenStream.getText(interval)}].")
  }
}
